_posts/2023-09-11-Rails-7-1-0-beta-1-has-been-released.markdown

```diff
@@ -0,0 +1,142 @@
+
+These files serve as a foundational setup for deploying your Rails application in a
+production environment using Docker.
+
+## Building blocks for authentication
+
+To complement with `has_secure_password` Rails 7.1 brings a few new building blocks to help developers to build their
+own authentication system.
+
+First, [`normalizes`](https://edgeapi.rubyonrails.org/classes/ActiveRecord/Normalization/ClassMethods.html#method-i-normalizes)
+declares an attribute normalization. This is useful to normalize attributes before saving them to the database.
+
+```ruby
+class User < ApplicationRecord
+  normalizes :email, with: -> email { email.strip.downcase }
+end
+
+user = User.create(email: " CRUISE-CONTROL@EXAMPLE.COM\n")
+user.email                  # => "cruise-control@example.com"
+```
+
+Second, [`authenticate_by`](https://edgeapi.rubyonrails.org/classes/ActiveRecord/SecurePassword/ClassMethods.html#method-i-authenticate_by)
+is intended to be used in the controller to authenticate the user, and it is implemented in a way that protect against
+common timing attacks.
+
+```ruby
+User.authenticate_by(email: "rafael@rubyonrails.org", password: "railsrocks")
+```
+
+[`generates_token_for`](https://edgeapi.rubyonrails.org/classes/ActiveRecord/TokenFor/ClassMethods.html#method-i-generates_token_for)
+can be used to implement features like password reset, email confirmation, and other features that require a single use
+token.
+
+```ruby
+class User < ApplicationRecord
+  has_secure_password
+
+  generates_token_for :password_reset, expires_in: 15.minutes do
+    # Last 10 characters of password salt, which changes when password is updated:
+    password_salt&.last(10)
+end
+
+user = User.first
+
+token = user.generate_token_for(:password_reset)
+User.find_by_token_for(:password_reset, token) # => user
+# 16 minutes later...
+User.find_by_token_for(:password_reset, token) # => nil
+```
+
+Finally, [`has_secure_password`](https://edgeapi.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html#method-i-has_secure_password)
+got the ability to automatically verify the current password when updating the password. This is useful to check
+if the user who is trying to update the password, knows the current password.
+
+```ruby
+# Schema: User(name:string, password_digest:string)
+class User < ActiveRecord::Base
+  has_secure_password
+end
+
+user = User.new(name: "rafael", password: "railsrocks", password_confirmation: "railsrocks")
+user.save                                                                      # => true
+user.update(password: "pwn3d", password_challenge: "")                         # => false, challenge doesn't authenticate
+user.update(password: "railsGetsEvenBetter", password_challenge: "railsrocks") # => true
+```
+
+## A more robust framework
+
+Rails promises to be a framework that can get you from Hello World to IPO, and on each release we are improving
```

Comment:
@dhh I tried to see if we wrote a blog post about the "Hello World to IPO" but I can only find your tweets. Do you have any post in your blog about this that we can link here?